server/src/main/java/org/elasticsearch/action/search/SearchRequestIntrospector.java

```diff
@@ -0,0 +1,245 @@
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.RangeQueryBuilder;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.sort.FieldSortBuilder;
+import org.elasticsearch.search.sort.ScoreSortBuilder;
+import org.elasticsearch.search.sort.SortBuilder;
+import org.elasticsearch.search.vectors.KnnVectorQueryBuilder;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Used to introspect a search request and extract metadata from it around the features it uses.
+ */
+public class SearchRequestIntrospector {
+
+    /**
+     * Introspects the provided search request and extracts metadata from it about some of its characteristics.
+     *
+     */
+    public static QueryMetadata introspectSearchRequest(SearchRequest searchRequest) {
+
+        String target = introspectIndices(searchRequest.indices());
+
+        String pitOrScroll = null;
+        if (searchRequest.scroll() != null) {
+            pitOrScroll = SCROLL;
+        }
+
+        SearchSourceBuilder searchSourceBuilder = searchRequest.source();
+        if (searchSourceBuilder == null) {
+            return new QueryMetadata(target, ScoreSortBuilder.NAME, HITS_ONLY, false, Strings.EMPTY_ARRAY, pitOrScroll);
+        }
+
+        if (searchSourceBuilder.pointInTimeBuilder() != null) {
+            pitOrScroll = PIT;
+        }
+
+        final String primarySort;
+        if (searchSourceBuilder.sorts() == null || searchSourceBuilder.sorts().isEmpty()) {
+            primarySort = ScoreSortBuilder.NAME;
+        } else {
+            primarySort = introspectPrimarySort(searchSourceBuilder.sorts().getFirst());
+        }
+
+        final String queryType = introspectQueryType(searchSourceBuilder);
+
+        QueryMetadataBuilder queryMetadataBuilder = new QueryMetadataBuilder();
+        if (searchSourceBuilder.query() != null) {
+            introspectQueryBuilder(searchSourceBuilder.query(), queryMetadataBuilder);
+        }
+
+        final boolean hasKnn = searchSourceBuilder.knnSearch().isEmpty() == false || queryMetadataBuilder.knnQuery;
+
+        return new QueryMetadata(
+            target,
+            primarySort,
+            queryType,
+            hasKnn,
+            queryMetadataBuilder.rangeFields.toArray(new String[0]),
+            pitOrScroll
+        );
+    }
+
+    private static final class QueryMetadataBuilder {
+        private boolean knnQuery = false;
+        private final List<String> rangeFields = new ArrayList<>();
+    }
+
+    public record QueryMetadata(
+        String target,
+        String primarySort,
+        String queryType,
+        boolean knn,
+        String[] rangeFields,
+        String pitOrScroll
+    ) {
+
+        public Map<String, Object> toAttributesMap() {
+            Map<String, Object> attributes = new HashMap<>();
```

Comment:
I tried to use constants for all strings as much as possible. No new strings are created per request. But the map is and I don't know how to get around that. I am a little worried about the performance hit caused by this.